yt/frontends/nc4_cm1/data_structures.py

```diff
@@ -0,0 +1,233 @@
+from yt.utilities.logger import ytLogger as mylog
+
+from .fields import CM1FieldInfo
+
+
+class CM1Grid(AMRGridPatch):
+    _id_offset = 0
+
+    def __init__(self, id, index, level, dimensions):
+        super(CM1Grid, self).__init__(id, filename=index.index_filename, index=index)
+        self.Parent = None
+        self.Children = []
+        self.Level = level
+        self.ActiveDimensions = dimensions
+
+    def __repr__(self):
+        return "CM1Grid_%04i (%s)" % (self.id, self.ActiveDimensions)
+
+
+class CM1Hierarchy(GridIndex):
+    grid = CM1Grid
+
+    def __init__(self, ds, dataset_type="cm1"):
+        self.dataset_type = dataset_type
+        self.dataset = weakref.proxy(ds)
+        # for now, the index file is the dataset!
+        self.index_filename = self.dataset.parameter_filename
+        self.directory = os.path.dirname(self.index_filename)
+        # float type for the simulation edges and must be float64 now
+        self.float_type = np.float64
+        super(CM1Hierarchy, self).__init__(ds, dataset_type)
+
+    def _initialize_state_variables(self):
+        super(CM1Hierarchy, self)._initialize_state_variables()
+        self.num_grids = 1
+
+    def _detect_output_fields(self):
+        # build list of on-disk fields for dataset_type 'cm1'
+        self.field_list = []
+        for vname in self.dataset.parameters["variable_names"]:
+            self.field_list.append(("cm1", vname))
+
+    def _count_grids(self):
+        # This needs to set self.num_grids
+        self.num_grids = 1
+
+    def _parse_index(self):
+        # This needs to fill the following arrays, where N is self.num_grids:
+        #   self.grid_left_edge         (N, 3) <= float64
+        #   self.grid_right_edge        (N, 3) <= float64
+        #   self.grid_dimensions        (N, 3) <= int
+        #   self.grid_particle_count    (N, 1) <= int
+        #   self.grid_levels            (N, 1) <= int
+        #   self.grids                  (N, 1) <= grid objects
+        #   self.max_level = self.grid_levels.max()
+        self.grid_left_edge[0][:] = self.ds.domain_left_edge[:]
+        self.grid_right_edge[0][:] = self.ds.domain_right_edge[:]
+        self.grid_dimensions[0][:] = self.ds.domain_dimensions[:]
+        self.grid_particle_count[0][0] = 0
+        self.grid_levels[0][0] = 1
+        self.max_level = 1
+
+    def _populate_grid_objects(self):
+        # For each grid g, this must call:
+        #   g._prepare_grid()
+        #   g._setup_dx()
+        # This must also set:
+        #   g.Children <= list of child grids
+        #   g.Parent   <= parent grid
+        # This is handled by the frontend because often the children must be
+        # identified.
+        self.grids = np.empty(self.num_grids, dtype="object")
+        for i in range(self.num_grids):
+            g = self.grid(i, self, self.grid_levels.flat[i], self.grid_dimensions[i])
+            g._prepare_grid()
+            g._setup_dx()
+            self.grids[i] = g
```

Comment:
@matthewturk can you weigh in here please ?

Reply:
So it depends.  The reason we ever do `self.grids[i]` is if we are creating an object array in advance.  I think it might also no longer be necessary -- there was a segfault that we ran into many years ago as a result of numpy object arrays, which is perhaps no longer operative.

But the idea was, create an empty numpy object array, then fill it.  It's possible that in AMRVAC it creates the grid objects elsewhere and they are already populated.